### PR TITLE
Add support for OpenIDConnect token refresh

### PR DESCRIPTION
**Pull request by pshchelo**
Creatively adapted from kubernetes-client/python-base.

Adds new setup extra 'oidc' to install required OAuth-related libraries.

Tested with local Keycloak installation.

### CODE REVIEW

**Comment by pshchelo** [comment]
should I add proper tests, with mocked out oauth dependencies?

**Comment by pshchelo** [comment]
rebased, black-ified, poetry-fied

**Comment by hjacobs** [comment]
@pshchelo sorry for the late response. In the meantime there was some refactoring to make the HTTP adapter more pluggable (#56). Can you adapt your PR accordingly?

Thanks!

**Comment by pshchelo** [comment]
Yes, sure, will do.
We'd really like to see this in pykube, as we have openid in our k8s, and usual python-kubernetes client is not for humans :-) and has its own bug with keycloak (PR is on review there but is stalled)

**Comment by hjacobs** [comment]
@pshchelo thanks!

**Comment by pshchelo** [comment]
rebased, conflicts resolved fixed. again, tested against our local keycloak, token is refreshed and the config file is updated.

@hjacobs plz review
Somehow I'm not sure how to properly update poetry.lock in such scenario. Only leave the libs I've added in the patch?

**Comment by hjacobs** [comment]
Sorry for the delayed response. I will have a look tomorrow.

File: pykube/config.py [hunk 1]
```diff
@@ -173,6 +173,12 @@ def users(self):
                     us[ur["name"]] = u = copy.deepcopy(ur["user"])
                     BytesOrFile.maybe_set(u, "client-certificate", self.kubeconfig_file)
                     BytesOrFile.maybe_set(u, "client-key", self.kubeconfig_file)
+                    if "auth-provider" in u:
+                        BytesOrFile.maybe_set(
+                            u["auth-provider"]["config"],
```
Comment (hjacobs) [inline pykube/config.py:205]: the config key might not be present (would raise KeyError)

**Comment by hjacobs** [comment]
@pshchelo can you provide me some instructions on how to test this locally with a test setup (e.g. Keycloak + kind or Minikube)?

File: pykube/http.py [hunk 2]
```diff
@@ -71,17 +80,96 @@ def _auth_gcp(self, request, token, expiry, config):
...
+                "missing dependencies for OIDC token refresh support "
+                "(try pip install pykube-ng[oidc]"
+            )
+        auth_config = config.user["auth-provider"]["config"]
+        if "idp-certificate-authority" in auth_config:
+            verify = auth_config["idp-certificate-authority"].filename()
+        else:
+            verify = None
+        # TODO add timeouts
+        response = requests.get(
```
Comment (hjacobs) [inline pykube/http.py]: please add a timeout like your TODO already suggests :smirk:

File: pykube/http.py [hunk 3]
```diff
@@ -71,17 +80,96 @@ def _auth_gcp(self, request, token, expiry, config):
...
+            f"{auth_config['idp-issuer-url']}/.well-known/openid-configuration",
+            verify=verify,
+        )
+
+        if response.status_code != HTTPStatus.OK:
```
Comment (hjacobs) [inline pykube/http.py]: if you want to catch all errors (e.g. connection problems), you would need to wrap it in `try`/`except` --- I assume you want to ignore failures and continue with the old token

**Comment by pshchelo** [comment]
> @pshchelo can you provide me some instructions on how to test this locally with a test setup (e.g. Keycloak + kind or Minikube)?

I'll try to come up with something, most probably with OpenID connect thru free tier of Google or Okta.
At my place we deploy k8s clusters pointing to a central internal keycloak instance, so I never deploy it myself :-/

**Comment by pshchelo** [comment]
@hjacobs added timeouts (using already configured default http one in the `http.py` module) and global ignore of any possible refresh errors, in which case I rely on retries and just re-use already present `id-token`.

your question on how to test this stuff eventually led to me writing a blog post on that

https://pshchelo.github.io/pykube-oidc-refresh.html

using dev Okta account for OpenID provider, microk8s snap on Ubuntu for simple Kubernetes cluster, and self-written script to get a refresh token (see the post).

If you have any questions, I'll be happy to try and answer :-)

**Comment by pshchelo** [comment]
@hjacobs ping ^

**Pull request merged**